File: binary-search/binary-search.cpp
#include <iostream>

using namespace std;

int BinarySearchIter(int array[], int n, int x);
int BinarySearchRecurs(int array[], int x, int indexStart, int indexEnd);

int main(){

	int n = 6;
	int array[] = {1, 2, 3, 4, 5, 6};

	cout << BinarySearchIter(array, n, 1) << endl;
	cout << BinarySearchIter(array, n, 6) << endl;
	cout << BinarySearchIter(array, n, 3) << endl;
	cout << BinarySearchIter(array, n, 7) << endl;
	cout << BinarySearchIter(array, n, 0) << endl;

	cout << BinarySearchRecurs(array, 1, 0, n - 1) << endl;
	cout << BinarySearchRecurs(array, 6, 0, n - 1) << endl;
	cout << BinarySearchRecurs(array, 3, 0, n - 1) << endl;
	cout << BinarySearchRecurs(array, 7, 0, n - 1) << endl;
	cout << BinarySearchRecurs(array, 0, 0, n - 1) << endl;

	return 0;
}


int BinarySearchIter(int array[], int n, int x){

	int indexStart = 0;
	int indexEnd = n - 1;

	// while subarray at least length 1
	while(indexStart <= indexEnd){

		int indexMiddle = indexStart + (indexEnd - indexStart) / 2;

		if(array[indexMiddle] < x){ indexStart = indexMiddle + 1; }
		else if(array[indexMiddle] > x){ indexEnd = indexMiddle - 1; }
		else{ return(indexMiddle); }
	}

	return(-1);
}


int BinarySearchRecurs(int array[], int x, int indexStart, int indexEnd){

	if(indexStart <= indexEnd){

		int indexMiddle = indexStart + (indexEnd - indexStart) / 2;

		if(array[indexMiddle] < x){ return(BinarySearchRecurs(array, x, indexMiddle + 1, indexEnd)); }
		else if(array[indexMiddle] > x){ return(BinarySearchRecurs(array, x, indexStart, indexMiddle - 1)); }
		else{ return(indexMiddle); }
	}

	else{
		return(-1);
	}

}

File: binary-search/binary-search.py
"""

Binary search

"""


def binary_search_recursive(x, query, index_start, index_end):
    if index_start <= index_end:
        index_middle = index_start + (index_end - index_start) / 2

        if x[index_middle] < query:
            return binary_search_recursive(x, query, index_middle + 1, index_end)
        elif x[index_middle] > query:
            return binary_search_recursive(x, query, index_start, index_middle - 1)
        else:
            return index_middle

    else:
        return None


def binary_search_iterative(x, query):
    index_start = 0
    index_end = len(x) - 1

    while index_start <= index_end:
        index_middle = index_start + (index_end - index_start) / 2

        if x[index_middle] < query:
            index_start = index_middle + 1
        elif x[index_middle] > query:
            index_end = index_middle - 1
        else:
            return index_middle

    return None


if __name__ == '__main__':
    x = range(10)
    print 'Sorted array: ' + str(x)
    print 'Contains 5: ' + str(binary_search_recursive(x, 5, 0, len(x) - 1))
    print 'Contains 11: ' + str(binary_search_recursive(x, 11, 0, len(x) - 1))
    print 'Contains 5: ' + str(binary_search_iterative(x, 5))
    print 'Contains 11: ' + str(binary_search_iterative(x, 11))


